cronjob_scripts/trigger-package-build.py

```diff
@@ -0,0 +1,217 @@
+        time.sleep(30)
+
+
+def get_build_os(target_arch: str) -> str:
+    try:
+        return TARGET_ARCH_TO_BUILD_OS[target_arch]
+    except KeyError:
+        raise ValueError(f"Unrecognised target arch: {target_arch}")
+
+
+def get_target_architectures() -> list[str]:
+    target_arch = os.environ.get("TARGET_ARCH", None)
+    if target_arch in TARGET_ARCH_TO_BUILD_OS:
+        return [target_arch]
+
+    if target_arch == "all":
+        return list(TARGET_ARCH_TO_BUILD_OS.keys())
+
+    try:
+        rustc_version_output = subprocess.run(
+            ["rustc", "--version", "--verbose"], capture_output=True, text=True
+        )
+        current_arch = [
+            line.split("host: ")[1]
+            for line in rustc_version_output.stdout.splitlines()
+            if line.startswith("host: ")
+        ][0]
+        return [current_arch]
+    except:
+        raise ValueError(f"rustc did not tell us its host: {rustc_version_output}")
+
+
+def get_popular_crates():
+    with open(
+        Path(__file__).resolve().parent.parent.joinpath("popular-crates.txt"), "r"
+    ) as file:
+        return [
+            line.strip()
+            for line in file
+            if line.strip() != "" and not line.startswith("#")
+        ]
+
+
+def get_excludes(tracking_worktree_path: str, days: int, max_failures: int):
+    """
+    if a crate has reached `max_failures` failures in the last `days` days then we exclude it
+    """
+    failures = Counter()
+
+    failure_filenames = list(Path(tracking_worktree_path).glob("*-*-*"))
+    failure_filenames.sort(reverse=True)
+
+    for failure_filename in failure_filenames[:days]:
+        with open(failure_filename, "r") as file:
+            failures.update(
+                Counter(line.strip() for line in file if line.strip() != "")
+            )
+
+    return [crate for crate, count in failures.items() if count >= max_failures]
+
+
+def get_repo_url() -> str:
+    repo_env_var = os.environ.get("GITHUB_REPOSITORY")
+    if repo_env_var:
+        return f"https://github.com/{repo_env_var}"
+    return subprocess.run(
+        ["gh", "repo", "view", "--json", "url", "--jq", ".url"],
+        capture_output=True,
+        text=True,
+    ).stdout.strip()
+
+
+def get_branch() -> str:
+    return subprocess.run(
+        ["git", "branch", "--show-current"],
+        capture_output=True,
+        text=True,
+    ).stdout.strip()
+
+
+def get_current_version_if_unbuilt(
+    repo_url: str,
+    crate: str,
+    target_arch: str,
+) -> CrateVersionDict | None:
+    try:
+        version = get_latest_version(crate)
+    except:
+        print(f"Failed to get latest version for {crate}")
+        return None
+
+    url = f"{repo_url}/releases/download/{crate}-{version['vers']}/{crate}-{version['vers']}-{target_arch}.tar.gz"
+    response = requests.head(url, allow_redirects=True)
+    if response.ok:
+        print(f"{url} already uploaded")
+        return None
+    # FIXME: handle rate limit errors?
```

Comment:
Given that we already get all tags from repo, I suppose we can use `git describe --match *`?

Reply:
Potentially we could, I guess. This would leave us with the case where we make a tag and then die before we can upload the tarball.

Potentially we could check crates.io + git tags for **all** crates, and pick a small random sample to check the github releases for every hour?

This is the kind of proposal that you just wouldn't dream of doing if the whole thing was written in bash, so I feel quite happy that I made this effort.

I'm not sure whether I want to do this in **this** pr. Maybe a fast follow up?

Reply:
> I'm not sure whether I want to do this in this pr. Maybe a fast follow up?

Yeah a follow-up is 👍

Reply:
Ah. I thought about this again, and I don't think we can use git tags:

We used to make tags like `cargo-benchcmp-0.4.5-aarch64-apple-darwin` and add a single tarball to it. We now make tags like `cargo-benchcmp-0.4.5` and add a tarball per arch.

This means that we can't just read the output of `git tag` to determine if we have built a thing for a particular arch, because it might have been built for a different architecture.